### DISCUSSION TITLE
Suggestion to make it easier to trust Deno's Node compatibility

### DISCUSSION THREAD

**Discussion by ffMathy** [denoland/deno#26745] [Ideas]
While officially you say that there's Node compatibility, some things are not implemented.

That's fine to me, but it makes it hard for me to decide on whether Deno is a good idea for us. 

I'd love to convert our whole codebase, but I am worried that I'll be stuck really late in the process due to incompatibility with some less known library that I need.

I gave Bun a shot. Was really surprised how much is missing compatibility-wise. A colleague even had a Blue Screen of Death while installing packages. They have a similar marketing strategy where they advertise compatibility.

I think this fear could be addressed, and that it could be huge for Deno. I really believe in Deno long term, and I want it to succeed.

# My suggestion
What if you started openly working on getting Node's test suite to pass using Deno?

You could even publish a page nightly with GitHub Pages or something similar, which could show the percentage of tests passing, so someone could track Node compatibility over time in a more accurate and confident manner.

Right now you do indeed display which features of which module isn't supported, but what's supported isn't necessarily one-to-one with what actually works. I used many Bun APIs for instance that they said were supported, but due to bugs, it ended up not actually working with particular libraries etc.

**Edit:** I made this project now, which displays compatibility daily: https://github.com/ffMathy/is-deno-compatible-yet

**Comment by bartlomieju** [comment]
Hey, thanks for raising this.

> That's fine to me, but it makes it hard for me to decide on whether Deno is a good idea for us.

You're not the only one and we are planning to address this more concretely this quarter.

> What if you started openly working on getting Node's test suite to pass using Deno?

This has been well underway - we are actually running some tests from Node.js test suite - you can see it here: https://github.com/denoland/deno/blob/main/tests/node_compat/config.jsonc. Admittedly, it's barely ~20% of all tests available in Node and we have some plans to improve it - mainly by running these tests out-of-band, in a separate repository so we can always run all tests and report which ones are actually broken and which ones are now passing. And obviously a frontend page will be added as well to keep track of the status based on Deno release.

> Right now you do indeed display which features of which module isn't supported, but what's supported isn't necessarily one-to-one with what actually works. I used many Bun APIs for instance that they said were supported, but due to bugs, it ended up not actually working with particular libraries etc.

We also have another project called `npm_smoke_tests` that runs various popular npm libraries under Deno and tests if they are working properly. This one is not currently public as it needs some more polish, but then again the plan is to make it open source and widely available.

**Reply by ffMathy** [reply]
This is amazing news. I was actually a bit depressed by the outlook of staying in Node forever.

Do you have an idea of how soon we will be able to track percentage of compatibility in an open manner on the front page?

**Reply by bartlomieju** [reply]
We want to address it (at least partially this quarter), but I can't give any guarantees at the moment. I will link back to a proper issue once there's some movement here.

**Reply by ffMathy** [reply]
Ah okay, so am I understanding it correctly that there is no issue tracking this currently?

Is there a Deno roadmap somewhere?

**Reply by bartlomieju** [reply]
That's correct, no public issue so far.

You can see current milestones for the midterm development here: https://github.com/denoland/deno/milestones

**Reply by ffMathy** [reply]
@bartlomieju I made this now: https://github.com/ffMathy/is-deno-compatible-yet

You can see it live here: https://ffmathy.github.io/is-deno-compatible-yet/

It is updated nightly automatically.

However, it is showing 82%, and you said 20%. Can you help me figure out why it's calculating wrong?

**Reply by kt3k** [reply]
https://ffmathy.github.io/is-deno-compatible-yet/

This page seems having only 645 filenames. Where do these come from?

We use [this script](https://github.com/denoland/deno/blob/1d49b3cb0f54eb8184acc00ec4bb3bd519653441/tests/node_compat/runner/setup.ts) for copying Node.js test cases from their repository, and that script detects 3681 files under `test/` directory of Node.js repo.

[This file](https://github.com/denoland/deno/blob/main/tests/node_compat/runner/TODO.md) lists what are remaining for us.

Note: we ignore [these directories](https://github.com/denoland/deno/blob/main/tests/node_compat/runner/setup.ts#L22-L41), and include all others.

**Reply by ffMathy** [reply]
I made big changes now. It should be calculating correctly from what I can see. I also made it so that you can show the individual tests and their state or collapse them.

It currently shows 7% compatibility. Is that correct now?

**Reply by kt3k** [reply]
Your list doesn't exclude `fixtures`, for example. The check [here](https://github.com/ffMathy/is-deno-compatible-yet/blob/65417168c165c56e37660b3035c43fb7c697f670/update.ts#L48) needs to be something like:

```js
!category || ignoredDirectories.some(dir => category.startsWith(dir))
```

because categories includes items like `fixtures/wpt/url`, `fixtures/es-modules/json-cache`, etc. On the other hand, `ignoredDirectories` list only includes `fixtures`.

Our counting is visible here: https://github.com/denoland/deno/blob/1d49b3cb0f54eb8184acc00ec4bb3bd519653441/tests/node_compat/runner/TODO.md

In our current calculation, there are 3681 files in total and 595 files are ported for now.

**Reply by ffMathy** [reply]
Made this PR: https://github.com/denoland/deno/pull/27189

With that merged, I would be able to truly piggyback on the same calculations as you are using.

**Reply by ffMathy** [reply]
Thank you for merging! I have now corrected the code to use the Deno code. I also converted the codebase from Bun to Deno 2 😄

**Comment by nhrones** [comment]
Wonderful idea!  Eventually could be modeled after the https://caniuse.com/ page.

**Reply by ffMathy** [reply]
I made this: https://ffmathy.github.io/is-deno-compatible-yet/

But PRs are welcome. It's open source. It updates nightly.

**Comment by axetroy** [comment]
Good point.

Users need to know how compatible Deno is with Node.

If there is a clear indicator showing more than 95% compatibility, it can greatly increase user confidence and motivate them to migrate to Deno.

**Reply by ffMathy** [reply]
I made this now: https://ffmathy.github.io/is-deno-compatible-yet/
